### PR TITLE
[1LP][WIP] Post upgrade test run support

### PR DESCRIPTION
**Pull request by digitronik**
## Purpose or Intent
- First step towards upgrade testing
- Adding `post-upgrade` support (for now it will support only `Minor` upgrade)
- Tested `5.11.0.25` upgrade to `5.11.z` l.e. latest
- Tested `5.10.10.0` upgrade to `5.10.z` l.e latest

For Reviewer: job `ndhandre-pr-9510`

### CODE REVIEW

File: cfme/fixtures/post_appliance_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        "--post-upgrade",
+        dest="post_upgrade",
+        action="store_true",
+        default=False,
+        help="Enable upgrade an appliance before the first test is run.",
+    )
+    group.addoption(
+        "--upgrade-to",
+        dest="upgrade_to",
+        action="store_true",
```
Comment (mshriver) [inline cfme/fixtures/post_appliance_upgrade.py]: I don't think this is the behavior you want here, you want the user to be able to pass a string.

File: cfme/fixtures/post_appliance_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import pytest
+
+
+def pytest_addoption(parser):
+    group = parser.getgroup("cfme")
+    group.addoption(
+        "--post-upgrade",
```
Comment (mshriver) [inline cfme/fixtures/post_appliance_upgrade.py]: I think this is a good name for the marker for tests themselves, but I'm thinking `--upgrade-appliance` would be a better CLI trigger for this. Then we have `--update-appliance` to trigger OS update, and `--upgrade-appliance` to trigger application upgrade

File: cfme/markers/upgrade.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    post_upgrade = config.getoption("upgrade_appliance")
+
+    if not post_upgrade:
+        return
+
+    keep, discard = [], []
+
+    for item in items:
+        # Filter test as per marker
+        if post_upgrade and item.get_marker("post_upgrade"):
```
Comment (mshriver) [inline cfme/markers/upgrade.py]: No need to check post_upgrade again, at this point the function should have returned if it was not set.

Comment (digitronik) [reply cfme/markers/upgrade.py]: right :+1:

File: cfme/markers/upgrade.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+# Filtering options
+def pytest_addoption(parser):
+    group = parser.getgroup("cfme")
+    group.addoption(
+        "--upgrade-appliance",
+        dest="upgrade_appliance",
+        action="store_true",
+        default=False,
+        help="Enable upgrade an appliance before the first test is run.",
```
Comment (john-dupuy) [inline cfme/markers/upgrade.py]:
```suggestion
        help="Upgrade an appliance before tests are run.",
```

File: cfme/utils/appliance/__init__.py [hunk 5]
```diff
@@ -1542,6 +1542,45 @@ def update_rhel(self, *urls, **kwargs):
 
         return result
 
+    def upgrade(self, upgrade_to, cfme_only=True, reboot=False):
+        """Upgrade an appliance"""
```
Comment (john-dupuy) [inline cfme/utils/appliance/__init__.py]: Can you add some information about upgrade paths? I.e. this is only for upgrading between z-streams (based off of L1550).

File: cfme/markers/upgrade.py (hunk 4 above)
Comment (digitronik) [reply cfme/markers/upgrade.py]: @john-dupuy I tried to fallow `update`. 
```
  --update-appliance    Enable updating an appliance before the first test is run.
```

**Comment by dajoRH** [comment]
Lint report for commit 237c241d0823858a002fcd5422ee53b93c297800:

`cfme/utils/appliance/__init__.py`:
- :heavy_exclamation_mark: Line 1556:1: **W293** *blank line contains whitespace*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/markers/upgrade.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        action="store",
+        default="5.11.z",
+        help="Supported versions 5.9.z, 5.10.z, 5.11.z (.z means latest and default is 5.11.z)",
+    )
+
+
+# Test collection modifier
+def pytest_collection_modifyitems(session, config, items):
+    """Provide filtering of test case collection based on the CLI options"""
+    post_upgrade = config.getoption("upgrade_appliance")
```
Comment (mshriver) [inline cfme/markers/upgrade.py]: I'm not sure this is the behavior we actually want.

There are two separate things being combined and made interdependent here:
1. post-upgrade pytest option, which should trigger the actual upgrade of appliances
2. the marker filtering of test collection

The first is obviously not provided by this function, we're just tying whether or not to modify collection based on it.  I don't think it should be linked. We should be able to run _all_ tests along with filter the collection of them.

You can just use `pytest -m post_upgrade` to filter test collection on the marker - without any of this code, its built into pytest `-m` option.  The only downside of using this is the uncollection metrics that get printed just include non-matched markers in the 'uncollectif' grouping.

```
pytest --collect-only -m junk cfme/tests/webui/test_general_ui.py

==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.7.4, pytest-3.4.1, py-1.8.0, pluggy-0.6.0
rootdir: /home/mshriver/repos/integration_tests, inifile:
plugins: polarion-collect-0.21.0, manageiq-integration-tests-17.61.1.dev9+g5b40aa1e1
collected 22 items                                                                                                                                                                           
Uncollection Stats:
 manual: 10
 uncollectif: 5
 0 tests left after all uncollections

```

And with a slight modification, adding a '@pytest.mark.junk' marker to a test function, with no other code changes:

```
pytest --collect-only -m junk cfme/tests/webui/test_general_ui.py

==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.7.4, pytest-3.4.1, py-1.8.0, pluggy-0.6.0
rootdir: /home/mshriver/repos/integration_tests, inifile:
plugins: polarion-collect-0.21.0, manageiq-integration-tests-17.61.1.dev9+g5b40aa1e1
collected 22 items                                                                                                                                                                           
<Module 'cfme/tests/webui/test_general_ui.py'>
  <Function 'test_add_provider_trailing_whitespaces[virtualcenter]'>
Uncollection Stats:
 manual: 10
 uncollectif: 5
 1 tests left after all uncollections

```

So the only value-add in defining the hookimpl that controls collection is that we can specify a particular uncollection_stat field for it, which I'm not sure we need.  The check against upgrade_appliance should not also impact whether or not the marked tests are included.  We should be able to: 1) collect post_upgrade tests without including the upgrade option and 2) include the upgrade option without limiting test collection

Comment (digitronik) [reply cfme/markers/upgrade.py]: @mshriver the behavior you want actually I removed https://github.com/ManageIQ/integration_tests/pull/9510/commits/164cc21f77f50f6b69e3ef1bd7c3b7f6c6c03cf2.

removing `modifier` and lets collect test with `-m`

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/markers/upgrade.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        "--upgrade-to",
+        dest="upgrade_to",
+        action="store",
+        default="5.11.z",
+        help="Supported versions 5.9.z, 5.10.z, 5.11.z (.z means latest and default is 5.11.z)",
+    )
+
+
+def pytest_sessionstart(session):
+    if store.parallelizer_role == "master":
```
Comment (izapolsk) [inline cfme/markers/upgrade.py]: we have to get rid of session scoped fixtures

File: cfme/utils/appliance/__init__.py [hunk 8]
```diff
@@ -1543,6 +1543,52 @@ def update_rhel(self, *urls, **kwargs):
...
+        update_url = conf.cfme_data["basic_info"][supported_stream_repo_map[upgrade_to]]
+
+        logger.debug("Adding update repo to appliance")
+        result = self.ssh_client.run_command(f"curl {update_url} -o /etc/yum.repos.d/update.repo")
+        logger.debug(result.output)
+
+        self.evmserverd.stop()
+
+        logger.info("Running yum update")
+        cmd = "cfme -y " if cfme_only else "-y"
```
Comment (izapolsk) [inline cfme/utils/appliance/__init__.py]: I guess you could add -y to yum command

Comment (digitronik) [reply cfme/utils/appliance/__init__.py]: It was copy paste from fixed miq command

File: cfme/markers/upgrade.py (hunk 7 above)
Comment (digitronik) [reply cfme/markers/upgrade.py]: I need to discuss with you... as per me we need session scope as we can not upgrade appliance for each test even not consider `module` scope as we are targeting multiple test from multiple file.

File: cfme/utils/appliance/__init__.py (hunk 8 above)
Comment (digitronik) [reply cfme/utils/appliance/__init__.py]: moved `-y` with yum

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/utils/appliance/__init__.py [hunk 9]
```diff
@@ -1543,6 +1543,52 @@ def update_rhel(self, *urls, **kwargs):
 
         return result
 
+    def upgrade(self, upgrade_to, cfme_only=True, reboot=False):
```
Comment (mshriver) [inline cfme/utils/appliance/__init__.py]: This looks like redefinition of the method on L1334 in master?

https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/appliance/__init__.py

**Comment by digitronik** [comment]
As part of clean up closing this.
I don't think, we need this enhancement in the current situation.

**Pull request closed without merging**